GEP-2907: TLS mode and allowed routes (#3190)
kubernetes-sigs/gateway-api: geps/gep-2907/index.md

The above diagram depicts these four segments as edges in a graph.

### TLS mode

TLS can be configured with two distinct modes:

* **Terminate**: the TLS connection is instantiated between the frontend and the

>   Gateway. The connection between the Gateway and the backend is either left unencrypted
>   or a new TLS connection between the two entities is instantiated.

Reviewer: This might be a little more clear:
```suggestion
  Gateway. The connection between the Gateway and the backend left unencrypted
 unless a new TLS connection between the two entities is configured via BackendTLSPolicy.
```

### Context

The `*Route` objects refer the Gateway (or a specific Listener of the Gateway) they
want to be attached to. A successful attachment can be granted by one of the two
following conditions:

* the `Listener.AllowedRoutes` field allows that specific Route `GroupKind` to be
  attached to it, or
* the `Listener.Protocol` field allows that specific Route to be attached to the
  Listener. This applies only in case the Gateway's AllowedRoutes field is not set.

In case the First condition is not satisfied (the Route references a Gateway that
does not explicitly allow such an attachment), the Route `Accepted` condition
is set to False with reason `NotAllowedByListeners`. In case the Route references
a Gateway with no Listener matching the protocol needed by the Route, the Route
`ResolvedRefs` condition is set to False with reason `UnsupportedProtocol`.

### What Routes are allowed by Listeners with TLS

The following is a summary of all the Routes and the TLS termination mode they support.

| Route | TLS with Terminate | TLS with Passthrough |
|-|-|-|
| `HTTPRoute` | yes | no |
| `GRPCRoute` | yes | no |
| `TLSRoute` | yes | yes |

> | `TCPRoute` | no | no |

Reviewer 2: We might need a note here that for TCPRoute, the packets are passed through untouched to the backend. If you want to terminate a TLS stream at a Gateway and forward a TCP stream unencrypted to the backend, use TLSRoute.

### Context

The `*Route` objects refer the Gateway (or a specific Listener of the Gateway) they
want to be attached to. A successful attachment can be granted by one of the two
following conditions:

* the `Listener.AllowedRoutes` field allows that specific Route `GroupKind` to be
  attached to it, or
* the `Listener.Protocol` field allows that specific Route to be attached to the
  Listener. This applies only in case the Gateway's AllowedRoutes field is not set.

In case the First condition is not satisfied (the Route references a Gateway that
does not explicitly allow such an attachment), the Route `Accepted` condition
is set to False with reason `NotAllowedByListeners`. In case the Route references
a Gateway with no Listener matching the protocol needed by the Route, the Route
`ResolvedRefs` condition is set to False with reason `UnsupportedProtocol`.

### What Routes are allowed by Listeners with TLS

The following is a summary of all the Routes and the TLS termination mode they support.

| Route | TLS with Terminate | TLS with Passthrough |
|-|-|-|
| `HTTPRoute` | yes | no |
| `GRPCRoute` | yes | no |
| `TLSRoute` | yes | yes |

> | `TCPRoute` | yes | yes |

Reviewer: E.g. it says `It is invalid to set this field if the Protocol field is "HTTP", "TCP", or "UDP".`

> | `TCPRoute` | yes | yes |

Author: It was based on a discussion we had about enabling TLS on TCPRoute. But I think that at least for the time being, even considering the high level of uncertainty the L4 routes have at the moment, we should be aligned with the current API specs, without aiming at improving it.
